### PR TITLE
[Console] add suggestions for debug commands: firewall, form, messenger, router

### PR DESCRIPTION
**Pull request by IonBazan**
| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | #43594
| License       | MIT
| Doc PR        | -

Adding Bash completion for following commands:

- `debug:firewall`
- `debug:form`
- `debug:messenger`
- `debug:router`

~Waiting for #43596 to be merged first as it adds testing utilities and `DescriptorHelper::getFormats()`.~

### CODE REVIEW

**Comment by TimoBakx** [comment]
The `DebugFirewallCommand` part looks good. 👍

File: src/Symfony/Component/Form/Tests/Command/DebugCommandTest.php [hunk 1]
```diff
@@ -186,6 +190,92 @@ class:%s
...
+            ['--format', ''],
+            ['txt', 'json'],
+        ];
+
+        yield 'form_type' => [
+            [''],
+            $this->getCoreTypes(),
+        ];
+
+        yield 'option for short name' => [
```
Comment (GromNaN) [inline src/Symfony/Component/Form/Tests/Command/DebugCommandTest.php]: You inverted descriptions "short name" vs "FQCN".

File: src/Symfony/Component/Form/Tests/Command/DebugCommandTest.php [hunk 2]
```diff
@@ -186,6 +190,92 @@ class:%s
...
+            ['Symfony\\Component\\Form\\Extension\\Core\Type\\ButtonType', ''],
```
Comment (GromNaN) [inline src/Symfony/Component/Form/Tests/Command/DebugCommandTest.php]: Is this double backslash missing?
```suggestion
            ['Symfony\\Component\\Form\\Extension\\Core\\Type\\ButtonType', ''],
```

**Review by GromNaN — CHANGES_REQUESTED** [review]
Very nice, especially on `debug:form`.

Namespace backslash are missing from completion. 

```console
$ bin/console debug:form 
AppFormCommentType                                     SymfonyComponentFormExtensionCoreTypeCheckboxType      SymfonyComponentFormExtensionCoreTypeEmailType         SymfonyComponentFormExtensionCoreTypePasswordType      SymfonyComponentFormExtensionCoreTypeTextType
AppFormPostType                                        SymfonyComponentFormExtensionCoreTypeChoiceType        SymfonyComponentFormExtensionCoreTypeFileType          SymfonyComponentFormExtensionCoreTypePercentType       SymfonyComponentFormExtensionCoreTypeTextareaType
AppFormTypeChangePasswordType                          SymfonyComponentFormExtensionCoreTypeCollectionType    SymfonyComponentFormExtensionCoreTypeFormType          SymfonyComponentFormExtensionCoreTypeRadioType         SymfonyComponentFormExtensionCoreTypeTimeType
AppFormTypeDateTimePickerType                          SymfonyComponentFormExtensionCoreTypeColorType         SymfonyComponentFormExtensionCoreTypeHiddenType        SymfonyComponentFormExtensionCoreTypeRangeType         SymfonyComponentFormExtensionCoreTypeTimezoneType
AppFormTypeTagsInputType                               SymfonyComponentFormExtensionCoreTypeCountryType       SymfonyComponentFormExtensionCoreTypeIntegerType       SymfonyComponentFormExtensionCoreTypeRepeatedType      SymfonyComponentFormExtensionCoreTypeUrlType
AppFormUserType                                        SymfonyComponentFormExtensionCoreTypeCurrencyType      SymfonyComponentFormExtensionCoreTypeLanguageType      SymfonyComponentFormExtensionCoreTypeResetType         SymfonyComponentFormExtensionCoreTypeWeekType
SymfonyBridgeDoctrineFormTypeEntityType                SymfonyComponentFormExtensionCoreTypeDateIntervalType  SymfonyComponentFormExtensionCoreTypeLocaleType        SymfonyComponentFormExtensionCoreTypeSearchType        
SymfonyComponentFormExtensionCoreTypeBirthdayType      SymfonyComponentFormExtensionCoreTypeDateTimeType      SymfonyComponentFormExtensionCoreTypeMoneyType         SymfonyComponentFormExtensionCoreTypeSubmitType        
SymfonyComponentFormExtensionCoreTypeButtonType        SymfonyComponentFormExtensionCoreTypeDateType          SymfonyComponentFormExtensionCoreTypeNumberType        SymfonyComponentFormExtensionCoreTypeTelType           
```

Option completion is very accurate with a class:
```console
$ bin/console debug:form App\Form\CommentType 
action                        block_name                    csrf_message                  disabled                      getter                        invalid_message               label_html                    priority                      trim                          
allow_extra_fields            block_prefix                  csrf_protection               empty_data                    help                          invalid_message_parameters    label_translation_parameters  property_path                 upload_max_size_message       
allow_file_upload             by_reference                  csrf_token_id                 error_bubbling                help_attr                     is_empty_callback             legacy_error_messages         required                      validation_groups             
attr                          compound                      csrf_token_manager            error_mapping                 help_html                     label                         mapped                        row_attr                      
attr_translation_parameters   constraints                   data                          extra_fields_message          help_translation_parameters   label_attr                    method                        setter                        
auto_initialize               csrf_field_name               data_class                    form_attr                     inherit_data                  label_format                  post_max_size_message         translation_domain            
```

But fails if backslashes are escaped. I don't know if that should be supported:
```console
$ bin/console debug:form App\\Form\\CommentType 


```
```
RuntimeException: The autoloader expected class "App\\Form\\CommentType" to be defined in file "src//Form//CommentType.php". The file was found but the class was not in it, the class name or namespace probably has a typo.
```

File: src/Symfony/Component/Form/Tests/Command/DebugCommandTest.php (hunk 2 above)
Comment (IonBazan) [reply src/Symfony/Component/Form/Tests/Command/DebugCommandTest.php]: Good catch, thanks!

**Comment by IonBazan** [comment]
@GromNaN these missing backslashes seem like an issue with the completion command escaping itself and input escaping. The suggestions provided by the `debug:form` do contain the backslashes so it must be something wrong with the terminal or the completion engine. Are there any similar issues in other commands too?

**Comment by GromNaN** [comment]
Same issue with backslashes in #43653.

**Comment by wouterj** [comment]
I've proposed a (quite ugly) fix for the backslashes: https://github.com/symfony/symfony/pull/43664 I welcome more testers of this script, as I'm not sure how well this actually works (it works on my machine™)

File: src/Symfony/Component/Form/Tests/Command/DebugCommandTest.php (hunk 1 above)
Comment (IonBazan) [reply src/Symfony/Component/Form/Tests/Command/DebugCommandTest.php]: Fixed 👍🏻

**Comment by IonBazan** [comment]
@GromNaN please review again

**Comment by GromNaN** [comment]
LGTM. I used this PR to test https://github.com/symfony/symfony/pull/43665 and it worked very well.

Regarding class types, the short names could have been completed in case someone type "ButtonType" for example. But name conflicts can appear and the completion made the FQCN  lot easier to type.

**Review by wouterj — APPROVED** [review]

File: src/Symfony/Component/Messenger/Tests/Command/DebugCommandTest.php [hunk 3]
```diff
@@ -166,4 +168,24 @@ public function testExceptionOnUnknownBusArgument()
         $tester = new CommandTester($command);
         $tester->execute(['bus' => 'unknown_bus'], ['decorated' => false]);
     }
+
+    /**
+     * @dataProvider provideCompletionSuggestions
+     */
+    public function testComplete(array $input, array $expectedSuggestions)
+    {
```
Comment (fabpot) [inline src/Symfony/Component/Messenger/Tests/Command/DebugCommandTest.php:176]: if (!class_exists(CommandCompletionTester::class)) {
            $this->markTestSkipped('Test command completion requires symfony/console 5.4+.');
        }

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Messenger/Tests/Command/DebugCommandTest.php (hunk 3 above)
Comment (IonBazan) [reply src/Symfony/Component/Messenger/Tests/Command/DebugCommandTest.php:176]: Resolved and squashed the commits.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @IonBazan.

**Pull request merged**